src/base.ts

```diff
@@ -21,6 +28,85 @@ export class MicrosoftConnector extends BaseHttpConnector<
     this.client = Client.initWithMiddleware(clientOptions)
   }
 
+  async onStart(): Promise<void> {
+    const logger = this.app.getLogger()
+    const subscriptions: MicrosoftGraph.Subscription[] = (
+      await this.sdk().api('/subscriptions').get()
+    ).value
+
+    for (const event of Object.values(this.eventConfigurations)) {
+      const options: MicrosoftEventConfigOptions = event.options
+      const existingSubscription = subscriptions.find((subscription) =>
+        Object.keys(event.options).every(
+          (key) => event.options[key] === subscription[key as keyof MicrosoftGraph.Subscription],
+        ),
+      )
+      let subscription: MicrosoftGraph.Subscription
+      if (existingSubscription) {
+        logger.info(
+          `Reshuffle Microsoft - existing webhook reused (resource: ${existingSubscription.resource}, url: ${existingSubscription.notificationUrl})`,
+        )
+        subscription = existingSubscription
+      } else {
+        subscription = await this.sdk().api('/subscriptions').post(options)
+        logger.info(
+          `Reshuffle Microsoft - webhook registered successfully (resource: ${subscription.resource}, url: ${subscription.notificationUrl})`,
+        )
+      }
+      this.eventConfigurations[event.id].options = {
+        ...this.eventConfigurations[event.id].options,
+        subscriptionId: subscription.id,
+      }
+    }
+  }
+
+  // Your events
+  on(
+    options: MicrosoftEventConfigOptions,
+    handler: (event: EventConfiguration & Record<string, any>, app: Reshuffle) => void,
+    eventId: string,
+  ): EventConfiguration {
+    const path = this.configOptions?.webhookPath || DEFAULT_WEBHOOK_PATH
+
+    if (!eventId) {
+      eventId = `Microsoft${path}/${options.resource}/${this.id}`
+    }
+    const event = new EventConfiguration(eventId, this, options)
+    this.eventConfigurations[event.id] = event
+
+    this.app.when(event, handler as any)
+    this.app.registerHTTPDelegate(path, this)
+
+    return event
+  }
+
+  async handle(req: Request, res: Response): Promise<boolean> {
+    const { validationToken } = req.query
+
+    if (validationToken) {
+      res.send(validationToken)
+    } else {
+      const data = req.body.value
+
+      for (const incomingEvent of data) {
+        const eventsUsingGithubEvent = Object.values(this.eventConfigurations).filter(
+          (event: EventConfiguration) =>
+            event.options.subscriptionId === incomingEvent.subsciptionId,
+        )
+
+        for (const event of eventsUsingGithubEvent) {
+          await this.app.handleEvent(event.id, {
+            ...event,
+            ...req.body,
+          })
+        }
+      }
+
+      res.status(202).send()
```

Comment:
Should this be sent before the `await`ing loop? How fast is Microsoft expecting the response to return?

Reply:
You have 30 seconds to respond. I think you only want to send a success response once the events have all been handled.

Reply:
But this is not a success response. It is 'accepted'. I think the meaning of it is exactly "I received your request and am now processing it"

Reply:
Yeah your right. I'll move it to before the requests are handled.